includes/Modules/Analytics_4/Report/Response.php

```diff
@@ -75,38 +83,19 @@ public function parse_response( Data_Request $data, $response ) {
 		$multiple_ranges = $ranges_count > 1;
 		$rows            = array();
 
-		foreach ( $date_ranges as $date_range ) {
-			$start = strtotime( $date_range->getStartDate() );
-			$end   = strtotime( $date_range->getEndDate() );
-
-			// Skip this date range if either start date or end date is corrupted.
-			if ( ! $start || ! $end ) {
-				continue;
-			}
-
-			// Loop through all days in the date range and check if there is a metric value
-			// for it. If the metric value is missing, we will need to add one with a zero value.
-			$now = $start;
-			do {
-				// Format the current time to a date string and add a day in seconds to the current date
-				// to shift to the next date.
-				$current_date = gmdate( 'Ymd', $now );
-				$now         += DAY_IN_SECONDS;
-
-				// Add rows for the current date for each date range.
+		// Add rows for the current date for each date range.
+		self::iterate_date_ranges(
+			$date_ranges,
+			function( $date ) use ( &$rows, $ranges_count, $metric_headers, $multiple_ranges ) {
```

Comment:
This `use` clause is missing `$existing_rows`, which means the existing row is never found, so we always return zero data for all rows. It's a small fix, I'll apply the change myself.

```suggestion
			function( $date ) use ( &$rows, $existing_rows, $ranges_count, $metric_headers, $multiple_ranges ) {
```

Reply:
Hi @eugene-manuilov - although I'd applied this fix myself, hoping to prevent an extra execution/review cycle, I've realised that the test coverage is lacking a check to ensure that the resulting rows which do have data retain this data and aren't overwritten with zeroes. Therefore, please can you update the tests to provide coverage for this aspect?